server/auth/range_perm_cache.go

```diff
@@ -105,6 +105,9 @@ func checkKeyPoint(lg *zap.Logger, cachedPerms *unifiedRangePermissions, key []b
 
 func (as *authStore) isRangeOpPermitted(tx AuthReadTx, userName string, key, rangeEnd []byte, permtyp authpb.Permission_Type) bool {
 	// assumption: tx is Lock()ed
+	as.rangePermCacheMu.Lock()
+	defer as.rangePermCacheMu.Unlock()
```

Comment:
Locking the whole method will definitely lead to decrease of performance. 

Firstly, I was thinking a sync.RWMutex could make more sense because usually the reading cache operations are greatly overnumber the writing cache operations. It means that we need to upgrade a read lock to a write lock when needed, but it complicates the design & implementation, and [golang](https://github.com/golang/go/issues/38859) doesn't support such case either :( .  

Secondly, [isOpPermitted](https://github.com/etcd-io/etcd/blob/81491914fbfbfcc041e175beb20f2cf3f575557e/server/auth/store.go#L839) is the only case which could be called by both the applying workflow and the API. All other cases ( I mean cases which need to access the `rangePermCache`) are coming from the applying workflow. Is it possible to resolve this from a higher level? 

An unrelated comment, the naming [isRangeOpPermitted](https://github.com/etcd-io/etcd/blob/81491914fbfbfcc041e175beb20f2cf3f575557e/server/auth/store.go#L874) isn't accurate, because it handles all operation types, including `authpb.READ` and `authpb.WRITE`.

Reply:
Another proposal: 
Firstly change the sync.Mutex to sync.RWMutex.

Secondly, the isRangeOpPermitted only reads the cache, and the cache is only updated in the following cases:

1. etcd startup & initialization
2. recovering from snapshot
3. RoleGrant and RoleRevoke

Does it make sense?

Reply:
@ahrtr I think your suggestion makes sense, it will simplify the design and implementation. Let me update this PR. Thanks!